L1Trigger/TrackFindingTracklet/plugins/ProducerKFout.cc

```diff
@@ -295,10 +293,12 @@ namespace trklet {
               digitise(TTTrack_TrackWord::chi2RZBins, tempchi2rz, (double)setup_->kfoutchi2rzConv())};
 
           // Run BDT emulation and package output into 3 bits
-
-          tempTQMVA = trackQualityModel_->runEmulatedTQ(trackQuality_inputs);
-          tempTQMVA = std::trunc(tempTQMVA * ap_fixed_rescale);
-          TTBV tqMVA(digitise(tqBins_, tempTQMVA, 1.0), TTTrack_TrackWord::TrackBitWidths::kMVAQualitySize, false);
+          // output needs sigmoid transformation applied
+          tempTQMVAPreSig = trackQualityModel_->runEmulatedTQ(trackQuality_inputs);
+          tempTQMVAPreSig = std::trunc(tempTQMVAPreSig * ap_fixed_rescale);
+          TTBV tqMVA(digitise(TTTrack_TrackWord::tqMVABins, 1. / (1. + exp(-tempTQMVAPreSig)), 1.0),
```

Comment:
Check that the ap_fixed_rescale multiplication isn't changing what's going on here, I think if you want to put a sigmoid on the output and not a inverse sigmoid on the bins you need to get rid of the ap_fixed_rescale multiplication but then we aren't matching emulation, this might not matter if the bin that is selected is the same anyway. The inverse sigmoid on the bins for defining this TTBV tqMVA might be better and then whenever we set the floating point trkMVA1 use a sigmoid on the output of the BDT.

Reply:
I implemented a function in TTTrack_Word to convert the 0-1 bins to the pre-sigmoid ones, and after testing it (and fixing a bug in the digitise function), I'm not sure the ap_fixed_rescale is working as expected. Here is a list of outputs for a couple of tracks:
```
post-sig value 0.992879, pre-sig value 4.9375, post-sig bin 7, pre-sig bin 7, pre-sig bin with rescale, 7
post-sig value 0.492188, pre-sig value -0.03125, post-sig bin 3, pre-sig bin 3, pre-sig bin with rescale, 2
post-sig value 0.972415, pre-sig value 3.5625, post-sig bin 7, pre-sig bin 7, pre-sig bin with rescale, 7
post-sig value 0.893309, pre-sig value 2.125, post-sig bin 7, pre-sig bin 7, pre-sig bin with rescale, 7
post-sig value 0.924142, pre-sig value 2.5, post-sig bin 7, pre-sig bin 7, pre-sig bin with rescale, 7
post-sig value 0.989672, pre-sig value 4.5625, post-sig bin 7, pre-sig bin 7, pre-sig bin with rescale, 7
post-sig value 0.384912, pre-sig value -0.46875, post-sig bin 3, pre-sig bin 3, pre-sig bin with rescale, 0
```
We can see the post-sig and pre-sig bins all match and are doing what is expected, and the rescale is multiplying the pre-sig value by 32.0 which sends the new value into the wrong bin when the pre-sig value is closer to 0. Why is this rescale happening? It seems to be causing quite a few binning mistakes and I don't think we would ever get anything within the 3-5 bins because of it.

Reply:
What bins do you have for the presigmoid case? What was the bug is the digitise function? Because when I do this on paper I get the same bins as I the post-sig version? Do you multiply your bins by the same ap_fixed rescale? The rescale is done because in the emulation we use the ap_fixed<10,5> datatype, in the firmware we use integers so to get from the ap_fixed type to integer we multiply by 2**(10 - 5) = 32. The result should be the same without the ap_fixed_rescale so long as the bins you are using are also without the ap_fixed_rescale but it makes comparing to the FW a lot easier.

Reply:
Ok so the issue must be that the pre-sigmoided bins were not rescaled. I will remove the rescaling as it seems a bit silly to multiply both the value we are comparing against and the bins by the same value, and that scaling can just be done easily when comparing with firmware. But I will keep the pre-sigmoided binning in the KFout. The bug in the digitise function was that the last bin is never being checked since the last bin is technically the last value in the array out to infinity. For the chi2 variables this bin is rarely filled but for the MVA the last bin is very often filled so perhaps that's why it was never caught before. Anyways, I'll make the changes and resolve this and then I think we should be good to go.